src/base/MiniV4Manager.sol

```diff
@@ -0,0 +1,237 @@
+        Position[] memory positions
+    ) private returns (BalanceDelta balanceDelta, BalanceDelta totalFeesAccrued) {
+        uint256 length = positions.length;
+
+        for (uint256 i; i != length; ++i) {
+            Position memory pos = positions[i];
+            IPoolManager.ModifyLiquidityParams memory params = IPoolManager.ModifyLiquidityParams({
+                tickLower: pos.tickLower,
+                tickUpper: pos.tickUpper,
+                liquidityDelta: -int128(pos.liquidity),
+                salt: pos.salt
+            });
+
+            (BalanceDelta delta, BalanceDelta feesAccrued) = poolManager.modifyLiquidity(poolKey, params, new bytes(0));
+            balanceDelta = balanceDelta + delta;
+            totalFeesAccrued = totalFeesAccrued + feesAccrued;
+        }
+    }
+
+    /**
+     * @dev Handles the collection of the fees during the `PoolManager` callback call
+     * @param poolKey Key of the Uniswap V4 pool to collect fees from
+     * @param positions Array of `Position` struct to collect fees from
+     * @return totalFees Fees collected from the positions
+     */
+    function _handleCollect(
+        PoolKey memory poolKey,
+        Position[] memory positions
+    ) private returns (BalanceDelta totalFees) {
+        uint256 length = positions.length;
+
+        for (uint256 i; i != length; ++i) {
+            Position memory pos = positions[i];
+            IPoolManager.ModifyLiquidityParams memory params = IPoolManager.ModifyLiquidityParams({
+                tickLower: pos.tickLower,
+                tickUpper: pos.tickUpper,
+                liquidityDelta: 0,
+                salt: pos.salt
+            });
+
+            (, BalanceDelta feesAccrued) = poolManager.modifyLiquidity(poolKey, params, new bytes(0));
+            totalFees = totalFees + feesAccrued;
+        }
+    }
+
+    /**
+     * @dev Handles the settlement of the balances during the `PoolManager` callback call
+     * @param poolKey Key of the Uniswap V4 pool, used to retrieve the currencies
+     * @param delta Current balances to settle denominated in `currency0` and `currency1`
+     */
+    function _handleSettle(PoolKey memory poolKey, BalanceDelta delta) private {
+        if (delta.amount0() > 0) {
+            poolManager.take(poolKey.currency0, address(this), delta.amount0().toUint128());
+        }
+
+        if (delta.amount1() > 0) {
+            poolManager.take(poolKey.currency1, address(this), delta.amount1().toUint128());
+        }
+
+        if (delta.amount0() < 0) {
+            poolManager.sync(poolKey.currency0);
+            poolKey.currency0.transfer(address(poolManager), uint256(-int256(delta.amount0())));
+            poolManager.settle();
+        }
+
+        if (delta.amount1() < 0) {
+            poolManager.sync(poolKey.currency1);
+            poolKey.currency1.transfer(address(poolManager), uint256(-int256(delta.amount1())));
+            poolManager.settle();
+        }
```

Comment:
Looks like this settlement logic doesn't work for native tokens as `settle` only considers the `msg.value` passed to be the amount to settle

https://github.com/Uniswap/v4-core/blob/59d3ecf53afa9264a16bba0e38f4c5d2231f80bc/src/PoolManager.sol#L353-L354